File: Python/projectLM.py
from abc import ABC, abstractmethod


class LibraryItem(ABC):
    def __init__(self, title, author, item_id):
        self.title = title
        self.author = author
        self.item_id = item_id
        self.__available = True        

    def is_available(self):
        return self.__available

    def borrow(self):
        if not self.__available:
            print(f" '{self.title}' is not available!")
            return False
        self.__available = False
        print(f" '{self.title}' borrowed successfully!")
        return True

    def return_item(self):
        self.__available = True
        print(f" '{self.title}' returned successfully!")

  
    def item_type(self):
        pass

    
    def display_info(self):
        pass


class Book(LibraryItem):
    def __init__(self, title, author, item_id, pages):
        super().__init__(title, author, item_id)
        self.pages = pages

    def item_type(self):
        return " Book"

    def display_info(self):
        status = "Available" if self.is_available() else " Borrowed"
        print(f"""
{self.item_type()}
   ID      : {self.item_id}
   Title   : {self.title}
   Author  : {self.author}
   Pages   : {self.pages}
   Status  : {status}
        """)


class Magazine(LibraryItem):
    def __init__(self, title, author, item_id, issue):
        super().__init__(title, author, item_id)
        self.issue = issue

    def item_type(self):
        return " Magazine"

    def display_info(self):
        status = " Available" if self.is_available() else " Borrowed"
        print(f"""
{self.item_type()}
   ID      : {self.item_id}
   Title   : {self.title}
   Author  : {self.author}
   Issue   : {self.issue}
   Status  : {status}
        """)


class DVD(LibraryItem):
    def __init__(self, title, author, item_id, duration):
        super().__init__(title, author, item_id)
        self.duration = duration

    def item_type(self):
        return " DVD"

    def display_info(self):
        status = " Available" if self.is_available() else " Borrowed"
        print(f"""
{self.item_type()}
   ID       : {self.item_id}
   Title    : {self.title}
   Director : {self.author}
   Duration : {self.duration} mins
   Status   : {status}
        """)



class Library:
    def __init__(self, name):
        self.name = name
        self.__items = []               

    def add_item(self, item):
        self.__items.append(item)
        print(f" Added '{item.title}' to library")

    def show_all(self):
        print(f"\n {self.name} — All Items:")
        print("=" * 40)
        for item in self.__items:
            item.display_info()         

    def search(self, title):
        print(f"\n Searching for: {title}")
        for item in self.__items:
            if title.lower() in item.title.lower():
                item.display_info()
                return item
        print(f" '{title}' not found!")
        return None

    def borrow_item(self, title):
        item = self.search(title)
        if item:
            item.borrow()

    def return_item(self, title):
        item = self.search(title)
        if item:
            item.return_item()

    def show_available(self):
        print(f"\n Available Items in {self.name}:")
        for item in self.__items:
            if item.is_available():
                print(f"   {item.item_type()} — {item.title}")


lib = Library("City Central Library")


lib.add_item(Book("Python Crash Course", "Eric Matthes", "B001", 560))
lib.add_item(Book("Clean Code", "Robert Martin", "B002", 431))
lib.add_item(Magazine("Tech Today", "John Doe", "M001", "March 2026"))
lib.add_item(DVD("The Social Network", "David Fincher", "D001", 120))


lib.show_all()

lib.borrow_item("Python Crash Course")
lib.borrow_item("Python Crash Course")   


lib.show_available()


lib.return_item("Python Crash Course")
lib.show_available()

File: Golang/projectLM.go
package main

import (
    "fmt"
    "strings"
)

type LibraryItem interface {
    ItemType() string
    DisplayInfo()
    Borrow() bool
    ReturnItem()
    IsAvailable() bool
}

type BaseItem struct {
    Title     string
    Author    string
    ItemID    string
    available bool    
}

func (b *BaseItem) IsAvailable() bool { return b.available }

func (b *BaseItem) Borrow() bool {
    if !b.available {
        fmt.Printf(" '%s' is not available!\n", b.Title)
        return false
    }
    b.available = false
    fmt.Printf(" '%s' borrowed successfully!\n", b.Title)
    return true
}

func (b *BaseItem) ReturnItem() {
    b.available = true
    fmt.Printf(" '%s' returned successfully!\n", b.Title)
}


type BOOK struct {
    BaseItem
    Pages int
}

type Magazine struct {
    BaseItem
    Issue string
}

type DVD struct {
    BaseItem
    Duration int
}


func (b BOOK) ItemType() string { return " Book" }
func (b BOOK) DisplayInfo() {
    status := " Available"
    if !b.IsAvailable() {
        status = " Borrowed"
    }
    fmt.Printf("\n%s | %s | by %s | %d pages | %s\n",
        b.ItemType(), b.Title, b.Author, b.Pages, status)
}


func (m Magazine) ItemType() string { return " Magazine" }
func (m Magazine) DisplayInfo() {
    status := " Available"
    if !m.IsAvailable() {
        status = " Borrowed"
    }
    fmt.Printf("\n%s | %s | Issue: %s | %s\n",
        m.ItemType(), m.Title, m.Issue, status)
}


func (d DVD) ItemType() string { return " DVD" }
func (d DVD) DisplayInfo() {
    status := " Available"
    if !d.IsAvailable() {
        status = " Borrowed"
    }
    fmt.Printf("\n%s | %s | %d mins | %s\n",
        d.ItemType(), d.Title, d.Duration, status)
}


type Library struct {
    Name  string
    items []LibraryItem  
}

func (l *Library) AddItem(item LibraryItem) {
    l.items = append(l.items, item)
}

func (l *Library) ShowAll() {
    fmt.Printf("\n %s — All Items:\n", l.Name)
    fmt.Println("=" * 40)
    for _, item := range l.items {
        item.DisplayInfo()              
    }
}

func (l *Library) Search(title string) LibraryItem {
    for _, item := range l.items {
        if strings.Contains(
            strings.ToLower(fmt.Sprintf("%v", item)),
            strings.ToLower(title)) {
            return item
        }
    }
    fmt.Printf(" '%s' not found!\n", title)
    return nil
}

func (l *Library) ShowAvailable() {
    fmt.Printf("\n Available Items in %s:\n", l.Name)
    for _, item := range l.items {
        if item.IsAvailable() {
            fmt.Printf("   %s\n", item.ItemType())
        }
    }
}

func PROJECTlM() {
    lib := Library{Name: "City Central Library"}

   
    B1 := &Book{baseitem: BaseItem{Title: "Python Crash Course",
        Author: "Eric Matthes", ItemID: "B001", available: true}, Pages: 560}
    B2 := &Book{BASEITEM: BaseItem{Title: "Clean Code",
        Author: "Robert Martin", ItemID: "B002", available: true}, Pages: 431}
    m1 := &Magazine{BASEITEM: BaseItem{Title: "Tech Today",
        Author: "John Doe", ItemID: "M001", available: true}, Issue: "March 2026"}
    d1 := &DVD{BaseItem: BASEITEM{Title: "The Social Network",
        Author: "David Fincher", ItemID: "D001", available: true}, Duration: 120}

    lib.AddItem(B1)
    lib.AddItem(B2)
    lib.AddItem(m1)
    lib.AddItem(d1)

  
    lib.ShowAll()

    B1.BORROW()
    B2.BORROW()     

    lib.ShowAvailable()

  
    B1.RETURNITEM()
    lib.ShowAvailable()
}